File: sample-projects/kotlin/src/main/kotlin/de/sots/cellarsandcentaurs/adapter/persistence/PersistedCreatures.kt
package de.sots.cellarsandcentaurs.adapter.persistence

import de.sots.cellarsandcentaurs.application.CreatureFacade
import de.sots.cellarsandcentaurs.domain.model.Creature
import de.sots.cellarsandcentaurs.domain.model.CreatureId
import de.sots.cellarsandcentaurs.domain.model.NoSuchCreatureException
import de.sots.cellarsandcentaurs.domain.service.Creatures
import org.slf4j.LoggerFactory

class PersistedCreatures(private val repository: CreatureRepository) : Creatures {
    private val logger = LoggerFactory.getLogger(PersistedCreatures::class.java)

    override fun save(creature: Creature) {
        logger.debug("Saving creature {}", creature.id)
        repository.save(CreatureEntity(creature.id.id))
    }

    override fun find(id: CreatureId): Creature {
        val creatureEntity = repository.findOne(id.id) ?: throw NoSuchCreatureException(id)
        return Creature(CreatureId(creatureEntity.id), CreatureFacade.STANDARD_CREATURE_TYPE)
    }
}

File: sample-projects/python/de/sots/cellarsandcentaurs/adapter/persistence/persisted_creatures.py
from de.sots.cellarsandcentaurs.application import CreatureFacade
from de.sots.cellarsandcentaurs.domain.model.creature import Creature
from de.sots.cellarsandcentaurs.domain.model.creature_id import CreatureId
from de.sots.cellarsandcentaurs.domain.model.errors import NoSuchCreatureException
from de.sots.cellarsandcentaurs.domain.service.creatures import Creatures

from .creature_entity import CreatureEntity
from .creature_repository import CreatureRepository


class PersistedCreatures(Creatures):
    def __init__(self, repository: CreatureRepository):
        self.repository = repository

    def save(self, creature: Creature) -> None:
        self.repository.save(CreatureEntity(creature.get_id().id))

    def find(self, creature_id: CreatureId) -> Creature:
        entity = self.repository.find_one(creature_id.id)
        if entity is None:
            raise NoSuchCreatureException(creature_id)
        return Creature(CreatureId(entity.get_id()), CreatureFacade.STANDARD_CREATURE_TYPE)
